In [ ]:
import pandas as pd
df=pd.read_csv('/content/final_reports_2016-23_cons_2024-12-24.csv', sep=';', quotechar='"')

In [ ]:
display(df.head())

NtsbNo EventType    Mkey             EventDate           City  \
0  ERA24LA084       ACC  193603  2023-12-31T15:13:00Z        Midland   
1  ERA24LA079       ACC  193587  2023-12-30T15:04:00Z  Daytona Beach   
2  CEN24LA081       ACC  193605  2023-12-29T15:27:00Z       Beaumont   
3  WPR24LA062       ACC  193584  2023-12-28T10:57:00Z    Lake Havasu   
4  WPR24LA061       ACC  193576  2023-12-26T13:21:00Z       Prescott   

        State        Country ReportNo                  N#  \
0    Virginia  United States      NaN               N37GA   
1     Florida  United States      NaN  N828AK, FA3XNWMRAN   
2       Texas  United States      NaN               N71MS   
3  California  United States      NaN              N312PR   
4     Arizona  United States      NaN               N29TT   

           SerialNumber  ...  WeatherCondition                 Operator  \
0                 004CE  ...               VMC                      NaN   
1  1689, 163DF81001N020  ...               VMC   Tunica Helicopters LLC   
2                 7452C  ...               VMC  TAF AERIAL SERVICES LLC   
3                  1127  ...               VMC                      NaN   
4                   451  ...               VMC                      NaN   

  BroadPhaseofFlight ReportStatus RepGenFlag MostRecentReportType  \
0            Landing    Completed        NaN                Final   
1           Approach    Completed        NaN                Final   
2            Landing    Completed        NaN                Final   
3            Landing    Completed        NaN                Final   
4            Landing    Completed        NaN                Final   

                                       DocketUrl  \
0  https://data.ntsb.gov/Docket?ProjectID=193603   
1                                            NaN   
2  https://data.ntsb.gov/Docket?ProjectID=193605   
3  https://data.ntsb.gov/Docket?ProjectID=193584   
4  https://data.ntsb.gov/Docket?ProjectID=193576   

                                           ReportUrl  \
0  https://data.ntsb.gov/carol-repgen/api/Aviatio...   
1  https://data.ntsb.gov/carol-repgen/api/Aviatio...   
2  https://data.ntsb.gov/carol-repgen/api/Aviatio...   
3  https://data.ntsb.gov/carol-repgen/api/Aviatio...   
4  https://data.ntsb.gov/carol-repgen/api/Aviatio...   

                                            rep_text  rep_num_jsg  
0  Page 1 of 5\nAviation Investigation Final Repo...   ERA24LA084  
1  Page 1 of 7\nAviation Investigation Final Repo...   ERA24LA079  
2  Page 1 of 5\nAviation Investigation Final Repo...   CEN24LA081  
3  Page 1 of 5\nAviation Investigation Final Repo...   WPR24LA062  
4  Page 1 of 8\nAviation Investigation Final Repo...   WPR24LA061  

[5 rows x 48 columns]

In [ ]:
from datasets import load_dataset

ds = load_dataset("elihoole/asrs-aviation-reports")

In [ ]:
# Convert the Hugging Face dataset 'ds' to a pandas DataFrame
ds_df = ds['train'].to_pandas()

# Display the first few rows of the new DataFrame
print("First 5 rows of the 'ds' dataset (as DataFrame):")
display(ds_df.head())

First 5 rows of the 'ds' dataset (as DataFrame):


acn_num_ACN Time_Date Time.1_Local Time Of Day Place_Locale Reference  \
0     1574675    201808                0601-1200            SNA.Airport   
1     1224894    201412                0601-1200            MSY.Airport   
2     1134202    201312                1201-1800              ZZZ.ARTCC   
3     1222074    201411                1201-1800            CWA.Airport   
4     1733019    202003                1801-2400              ZDV.ARTCC   

  Place.1_State Reference Place.2_Relative Position.Angle.Radial  \
0                      CA                                          
1                      LA                                          
2                      US                                          
3                      WI                                          
4                      CO                                          

  Place.3_Relative Position.Distance.Nautical Miles  \
0                                                     
1                                                     
2                                                     
3                                              20.0   
4                                                     

  Place.4_Altitude.AGL.Single Value Place.5_Altitude.MSL.Single Value  \
0                                                              5000.0   
1                            1000.0                                     
2                                                              2600.0   
3                                                              4000.0   
4                                                             32000.0   

  Environment_Flight Conditions  ...  \
0                                ...   
1                           VMC  ...   
2                           IMC  ...   
3                                ...   
4                                ...   

                                Events.3_Detector Events.4_When Detected  \
0  Person Air Traffic Control; Person Flight Crew              In-flight   
1                              Person Flight Crew              In-flight   
2                              Person Flight Crew              In-flight   
3                              Person Flight Crew              In-flight   
4                              Person Flight Crew              In-flight   

                                     Events.5_Result  \
0  Air Traffic Control Issued New Clearance; Flig...   
1                      General None Reported / Taken   
2  Air Traffic Control Provided Assistance; Air T...   
3  Flight Crew Became Reoriented; General Mainten...   
4  Air Traffic Control Issued New Clearance; Flig...   

  Assessments_Contributing Factors / Situations Assessments.1_Primary Problem  \
0                                 Human Factors                 Human Factors   
1                                 Human Factors                 Human Factors   
2   Human Factors; Aircraft; Procedure; Weather                      Aircraft   
3                                      Aircraft                      Aircraft   
4                   Airspace Structure; Weather                       Weather   

                                  Report 1_Narrative Report 1.1_Callback  \
0  SNA RNP-Z to Runway 20R. The FMC was properly ...                       
1  On base to final turn to runway 1 in MSY at ap...                       
2  I climbed to my filed altitude of 5;000 FT; an...                       
3  I had my pitot heat checked prior to winter an...                       
4  At 32000 ft. just north of PUB the aircraft ex...                       

                                  Report 2_Narrative Report 2.1_Callback  \
0  We were cleared for the RNP RNAV Z 20R Approac...                       
1                                                                          
2                                                                          
3                                                                          
4            

In [ ]:
# Display the first few rows of the 'df' dataset for comparison
print("\nFirst 5 rows of the 'df' dataset:")
display(df.head())

Combining the two datasets


In [ ]:
# Add a 'source' column to each DataFrame
df['source'] = 'NTSB'
ds_df['source'] = 'ASRS'

In [ ]:
# Define column renaming maps for better consistency
# Renaming for df (NTSB data)
nts_df_rename_map = {
    'NtsbNo': 'ReportID',
    'EventDate': 'EventDate',
    'Make': 'AircraftMake',
    'Model': 'AircraftModel',
    'EventType': 'EventType',
    'ProbableCause': 'Narrative',
    'Findings': 'FindingsDetail',
    'AirCraftCategory': 'AircraftCategory',
    'HighestInjuryLevel': 'HighestInjuryLevel',
    'FatalInjuryCount': 'FatalInjuryCount',
    'SeriousInjuryCount': 'SeriousInjuryCount',
    'MinorInjuryCount': 'MinorInjuryCount',
    'OnboardInjuryCount': 'OnboardInjuryCount',
    'OnGroundInjuryCount': 'OnGroundInjuryCount'
}

# Renaming for ds_df (ASRS data) - based on common ASRS column names
# Please note: these are assumed common names. If ds_df has different names,
# this map will need adjustment based on the actual columns in ds_df.
asrs_df_rename_map = {
    'report_number': 'ReportID', # Assuming 'report_number' exists in ASRS data
    'event_date': 'EventDate', # Assuming 'event_date' exists in ASRS data
    'aircraft_type': 'AircraftModel', # ASRS often has aircraft type, we map to model for simplicity
    'aircraft': 'AircraftMake', # If specific make is available
    'narrative_text': 'Narrative', # Common ASRS field for incident description
    'narrative': 'Narrative',
    'primary_problem_area': 'EventType', # Can be mapped to event type
    'highest_injury_level': 'HighestInjuryLevel',
    'fatal_injury_count': 'FatalInjuryCount',
    'serious_injury_count': 'SeriousInjuryCount',
    'minor_injury_count': 'MinorInjuryCount',
    'onboard_injury_count': 'OnboardInjuryCount',
    'on_ground_injury_count': 'OnGroundInjuryCount'
}

# Apply renaming, ignore errors for columns not found in a specific dataframe
df_renamed = df.rename(columns=nts_df_rename_map, errors='ignore')
ds_df_renamed = ds_df.rename(columns=asrs_df_rename_map, errors='ignore')

In [ ]:
# Concatenate the two dataframes. Non-matching columns will be filled with NaN.
combined_df = pd.concat([df_renamed, ds_df_renamed], ignore_index=True)

print("Combined DataFrame head:")
display(combined_df.head())
print(f"\nShape of combined DataFrame: {combined_df.shape}")
print("\nColumns in combined DataFrame:")
print(combined_df.columns.tolist())

Combined DataFrame head:


ReportID EventType      Mkey             EventDate           City  \
0  ERA24LA084       ACC  193603.0  2023-12-31T15:13:00Z        Midland   
1  ERA24LA079       ACC  193587.0  2023-12-30T15:04:00Z  Daytona Beach   
2  CEN24LA081       ACC  193605.0  2023-12-29T15:27:00Z       Beaumont   
3  WPR24LA062       ACC  193584.0  2023-12-28T10:57:00Z    Lake Havasu   
4  WPR24LA061       ACC  193576.0  2023-12-26T13:21:00Z       Prescott   

        State        Country ReportNo                  N#  \
0    Virginia  United States      NaN               N37GA   
1     Florida  United States      NaN  N828AK, FA3XNWMRAN   
2       Texas  United States      NaN               N71MS   
3  California  United States      NaN              N312PR   
4     Arizona  United States      NaN               N29TT   

           SerialNumber  ... Events.3_Detector Events.4_When Detected  \
0                 004CE  ...               NaN                    NaN   
1  1689, 163DF81001N020  ...               NaN                    NaN   
2                 7452C  ...               NaN                    NaN   
3                  1127  ...               NaN                    NaN   
4                   451  ...               NaN                    NaN   

  Events.5_Result Assessments_Contributing Factors / Situations  \
0             NaN                                           NaN   
1             NaN                                           NaN   
2             NaN                                           NaN   
3             NaN                                           NaN   
4             NaN                                           NaN   

  Assessments.1_Primary Problem Report 1_Narrative  Report 1.1_Callback  \
0                           NaN                NaN                  NaN   
1                           NaN                NaN                  NaN   
2                           NaN                NaN                  NaN   
3                           NaN                NaN                  NaN   
4                           NaN                NaN                  NaN   

   Report 2_Narrative  Report 2.1_Callback  Report 1.2_Synopsis  
0                 NaN                  NaN                  NaN  
1                 NaN                  NaN                  NaN  
2                 NaN                  NaN                  NaN  
3                 NaN                  NaN                  NaN  
4                 NaN                  NaN                  NaN  

[5 rows x 159 columns]


Shape of combined DataFrame: (46117, 159)

Columns in combined DataFrame:
['ReportID', 'EventType', 'Mkey', 'EventDate', 'City', 'State', 'Country', 'ReportNo', 'N#', 'SerialNumber', 'HasSafetyRec', 'Mode', 'ReportType', 'OriginalPublishedDate', 'DocketOriginalPublishedDate', 'HighestInjuryLevel', 'FatalInjuryCount', 'SeriousInjuryCount', 'MinorInjuryCount', 'OnboardInjuryCount', 'OnGroundInjuryCount', 'Narrative', 'FindingsDetail', 'EventID', 'Latitude', 'Longitude', 'AircraftMake', 'AircraftModel', 'AircraftCategory', 'AirportID', 'AirportName', 'AmateurBuilt', 'NumberOfEngines', 'EngineType', 'Scheduled', 'PurposeOfFlight', 'FAR', 'AirCraftDamage', 'WeatherCondition', 'Operator', 'BroadPhaseofFlight', 'ReportStatus', 'RepGenFlag', 'MostRecentReportType', 'DocketUrl', 'ReportUrl', 'rep_text', 'rep_num_jsg', 'acn_num_ACN', 'Time_Date', 'Time.1_Local Time Of Day', 'Place_Locale Reference', 'Place.1_State Reference', 'Place.2_Relative Position.Angle.Radial', 'Place.3_Relative Position.

In [ ]:
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46117 entries, 0 to 46116
Columns: 159 entries, ReportID to Report 1.2_Synopsis
dtypes: float64(10), object(149)
memory usage: 55.9+ MB


Please note: The column renaming for the `ds_df` (ASRS) dataset was based on common ASRS field names. If the actual column names in `ds_df` are different, the `asrs_df_rename_map` in the code above will need to be adjusted accordingly. You can inspect `ds_df.columns` if you encounter unexpected `NaN` values for supposedly common fields.

In [ ]:
print("Summary statistics for numerical columns:")
display(combined_df.describe())

print("\nSummary statistics for object (non-numerical) columns:")
display(combined_df.describe(include='object'))

In [ ]:
# Convert 'EventDate' column to datetime, coercing errors
combined_df['EventDate'] = pd.to_datetime(combined_df['EventDate'], errors='coerce')

# Display the info again to confirm the change
print("DataFrame info after converting 'EventDate':")
combined_df.info()

DataFrame info after converting 'EventDate':
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46117 entries, 0 to 46116
Columns: 159 entries, ReportID to Report 1.2_Synopsis
dtypes: datetime64[ns, UTC](1), float64(10), object(148)
memory usage: 55.9+ MB


In [ ]:
# Calculate the percentage of missing values for each column
missing_percentage = combined_df.isnull().sum() / len(combined_df) * 100

# Display columns with missing values, sorted in descending order
missing_percentage = missing_percentage[missing_percentage > 0].sort_values(ascending=False)

print("Percentage of missing values per column (only showing columns with missing values):")
display(pd.DataFrame(missing_percentage, columns=['Missing Percentage']))

In [ ]:
# Calculate the percentage of missing values for each column again
missing_percentage = combined_df.isnull().sum() / len(combined_df) * 100

# Display columns with missing values, sorted in descending order
missing_percentage = missing_percentage[missing_percentage > 0].sort_values(ascending=False)

print("Percentage of missing values per column (only showing columns with missing values):")
display(pd.DataFrame(missing_percentage, columns=['Missing Percentage']))

In [ ]:
print("Summary statistics for numerical columns:")
display(combined_df.describe())

print("\nSummary statistics for object (non-numerical) columns:")
display(combined_df.describe(include='object'))

In [ ]:
# Identify columns with more than 90% missing values
columns_to_drop_high_missing = missing_percentage[missing_percentage > 90].index.tolist()

# Drop these columns from combined_df
combined_df = combined_df.drop(columns=columns_to_drop_high_missing)

print(f"Dropped {len(columns_to_drop_high_missing)} columns with more than 90% missing values.")
print("Dropped columns:")
for col in columns_to_drop_high_missing:
    print(f"- {col}")

# Display the info again to confirm the columns are dropped
print("\nDataFrame info after dropping high missing columns:")
combined_df.info()

In [ ]:
print("Missing value percentages for NTSB data (source='NTSB'):")
ntsb_missing_percentage = combined_df[combined_df['source'] == 'NTSB'].isnull().sum() / len(combined_df[combined_df['source'] == 'NTSB']) * 100
ntsb_missing_percentage = ntsb_missing_percentage[ntsb_missing_percentage > 0].sort_values(ascending=False)
display(pd.DataFrame(ntsb_missing_percentage, columns=['Missing Percentage']))

print("\nMissing value percentages for ASRS data (source='ASRS'):")
asrs_missing_percentage = combined_df[combined_df['source'] == 'ASRS'].isnull().sum() / len(combined_df[combined_df['source'] == 'ASRS']) * 100
asrs_missing_percentage = asrs_missing_percentage[asrs_missing_percentage > 0].sort_values(ascending=False)
display(pd.DataFrame(asrs_missing_percentage, columns=['Missing Percentage']))

In [ ]:
# Calculate the percentage of missing values for each column again
missing_percentage = combined_df.isnull().sum() / len(combined_df) * 100

# Display columns with missing values, sorted in descending order
missing_percentage = missing_percentage[missing_percentage > 0].sort_values(ascending=False)

print("Percentage of missing values per column (only showing columns with missing values):")
display(pd.DataFrame(missing_percentage, columns=['Missing Percentage']))

In [ ]:
# Identify columns with 100% missing values
columns_to_drop = missing_percentage[missing_percentage == 100].index.tolist()

# Drop these columns from combined_df
combined_df = combined_df.drop(columns=columns_to_drop)

print(f"Dropped {len(columns_to_drop)} columns with 100% missing values: {columns_to_drop}")

# Display the info again to confirm the columns are dropped
print("\nDataFrame info after dropping 100% missing columns:")
combined_df.info()

# Task
Clean the `combined_df` DataFrame by:
1. Identifying and cleaning numerical outliers in 'Latitude', 'Longitude', 'FatalInjuryCount', 'SeriousInjuryCount', 'MinorInjuryCount', 'OnboardInjuryCount', and 'OnGroundInjuryCount' columns.
2. Imputing missing categorical values in object-type columns by replacing `NaN` with 'Unknown'.
3. Creating new derived fields: `total_injuries` (sum of all injury counts) and `severity_category` (categorizing events based on `total_injuries` into 'Low', 'Medium', or 'High' severity).
4. Re-evaluating and displaying the percentage of missing values after these cleaning steps.
Finally, provide a summary of all data cleaning and feature engineering steps performed.

## Identify and Clean Numerical Outliers

### Subtask:
Examine the distributions of 'Latitude', 'Longitude', and various injury count columns ('FatalInjuryCount', 'SeriousInjuryCount', 'MinorInjuryCount', 'OnboardInjuryCount', 'OnGroundInjuryCount'). Identify and clean any clear outliers that don't make sense by either capping them or replacing them with `NaN` if they are clearly erroneous values outside of plausible ranges.


**Reasoning**:
First, replace values in the 'Latitude' column that are outside the plausible range of -90 to 90 with NaN. This is a common data cleaning step for geographical coordinates.



In [ ]:
combined_df['Latitude'] = combined_df['Latitude'].apply(lambda x: x if -90 <= x <= 90 else pd.NA)
print("Cleaned 'Latitude' column for values outside [-90, 90].")

**Reasoning**:
Next,clean the 'Longitude' column by replacing values outside the plausible range of -180 to 180 with NaN, similar to the 'Latitude' cleaning.



In [ ]:
combined_df['Longitude'] = combined_df['Longitude'].apply(lambda x: x if -180 <= x <= 180 else pd.NA)
print("Cleaned 'Longitude' column for values outside [-180, 180].")

**Reasoning**:
Now,replace any negative values in the injury count columns ('FatalInjuryCount', 'SeriousInjuryCount', 'MinorInjuryCount', 'OnboardInjuryCount', 'OnGroundInjuryCount') with NaN, as injury counts cannot be negative.



In [ ]:
injury_columns = ['FatalInjuryCount', 'SeriousInjuryCount', 'MinorInjuryCount', 'OnboardInjuryCount', 'OnGroundInjuryCount']

for col in injury_columns:
    # Ensure the column is numeric first to handle potential non-numeric entries gracefully
    combined_df[col] = pd.to_numeric(combined_df[col], errors='coerce')
    # Replace negative values with NaN
    combined_df[col] = combined_df[col].apply(lambda x: x if x >= 0 else pd.NA)

print("Cleaned injury count columns by replacing negative values with NaN.")

**Reasoning**:
To identify any extremely high, implausible values (outliers) in the injury count columns, examine their descriptive statistics, specifically the maximum values. This will help determine if any capping or replacement with NaN is necessary for values that are clearly erroneous data entries.



In [ ]:
print("Descriptive statistics for injury count columns after cleaning for negative values:")
display(combined_df[injury_columns].describe())

**Reasoning**:
The previous `describe()` command for injury count columns did not yield numerical statistics (like `max`) because the columns were likely treated as object type due to `pd.NA` values. To properly examine numerical outliers,explicitly convert these columns to a numeric type (like `float`) before calculating descriptive statistics. This will ensure that `describe()` provides the full range of numerical summaries including the maximum values, which are crucial for identifying implausible outliers.



## Identify and Flag Numerical Outliers

### Subtask:
Create a new boolean column `is_extreme_injury` to flag rows where 'MinorInjuryCount' or 'OnboardInjuryCount' exceed a threshold of 50, without altering the original injury count values.


In [ ]:
injury_columns = ['FatalInjuryCount', 'SeriousInjuryCount', 'MinorInjuryCount', 'OnboardInjuryCount', 'OnGroundInjuryCount']

for col in injury_columns:
    # Convert to float, replacing any non-numeric values (including pd.NA) with NaN
    combined_df[col] = pd.to_numeric(combined_df[col], errors='coerce')

print("Descriptive statistics for injury count columns after ensuring numeric type:")
display(combined_df[injury_columns].describe())

In [ ]:
extreme_injury_threshold = 50

combined_df['is_extreme_injury'] = (
    (combined_df['MinorInjuryCount'] > extreme_injury_threshold) |
    (combined_df['OnboardInjuryCount'] > extreme_injury_threshold)
)

print(f"New column 'is_extreme_injury' created. Flagging cases where 'MinorInjuryCount' or 'OnboardInjuryCount' exceed {extreme_injury_threshold}.")

print("First 5 rows with 'MinorInjuryCount', 'OnboardInjuryCount', and 'is_extreme_injury':")
display(combined_df[['MinorInjuryCount', 'OnboardInjuryCount', 'is_extreme_injury']].head())

extreme_injury_count = combined_df['is_extreme_injury'].sum()
print(f"Total identified extreme injury cases: {extreme_injury_count}")

**Reasoning**:
Now that numerical outliers have been addressed for Latitude, Longitude, and injury counts, the next step is to impute missing categorical values in object-type columns as per the overall task requirements. I will iterate through all object-type columns and replace any NaN values with the string 'Unknown'.



In [ ]:
print("Imputing missing categorical values with 'Unknown'...")

# Get a list of all object (string) columns
object_columns = combined_df.select_dtypes(include='object').columns

# Impute missing values in these columns with 'Unknown'
for col in object_columns:
    combined_df[col] = combined_df[col].fillna('Unknown')

print("Missing categorical values imputed.")

# Verify imputation by checking missing values in object columns
print("\nMissing values in object columns after imputation:")
missing_in_object_columns = combined_df[object_columns].isnull().sum()
missing_in_object_columns = missing_in_object_columns[missing_in_object_columns > 0]
display(missing_in_object_columns)

 next step: to create the derived field `total_injuries` by summing up all the individual injury count columns. This will provide a consolidated measure of injury severity for each event.



In [ ]:
print("Creating 'total_injuries' column...")

injury_columns = ['FatalInjuryCount', 'SeriousInjuryCount', 'MinorInjuryCount', 'OnboardInjuryCount', 'OnGroundInjuryCount']

# Ensure all injury columns are numeric before summing
for col in injury_columns:
    combined_df[col] = pd.to_numeric(combined_df[col], errors='coerce')

# Sum the injury columns, treating NaN as 0 for the sum
combined_df['total_injuries'] = combined_df[injury_columns].fillna(0).sum(axis=1)

print("'total_injuries' column created.")
print("First 5 rows with 'total_injuries' column:")
display(combined_df[['FatalInjuryCount', 'SeriousInjuryCount', 'MinorInjuryCount', 'OnboardInjuryCount', 'OnGroundInjuryCount', 'total_injuries']].head())

**Reasoning**:
The previous code block successfully created the 'total_injuries' column but generated a `PerformanceWarning` due to DataFrame fragmentation. To address this and prevent future warnings, I will explicitly de-fragment `combined_df` by making a copy of it. Then, I will proceed to create the `severity_category` column based on the `total_injuries`, as per the overall task instructions.



In [ ]:
print("Creating 'total_injuries' and 'severity_category' columns...")

injury_columns = ['FatalInjuryCount', 'SeriousInjuryCount', 'MinorInjuryCount', 'OnboardInjuryCount', 'OnGroundInjuryCount']

# De-fragment the DataFrame to avoid PerformanceWarning
combined_df = combined_df.copy()

# Ensure all injury columns are numeric before summing
for col in injury_columns:
    combined_df[col] = pd.to_numeric(combined_df[col], errors='coerce')

# Sum the injury columns, treating NaN as 0 for the sum
combined_df['total_injuries'] = combined_df[injury_columns].fillna(0).sum(axis=1)

# Define bins and labels for severity categorization
bins = [-1, 0, 5, combined_df['total_injuries'].max() + 1] # -1 to include 0 injuries
labels = ['Low', 'Medium', 'High']

# Create 'severity_category' column
combined_df['severity_category'] = pd.cut(
    combined_df['total_injuries'],
    bins=bins,
    labels=labels,
    right=True # bins are (min, max]
)

print("'total_injuries' and 'severity_category' columns created.")
print("First 5 rows with new columns:")
display(combined_df[['total_injuries', 'severity_category']].head())
print("Value counts for 'severity_category':")
display(combined_df['severity_category'].value_counts())

**Reasoning**:
 re-evaluate and display the percentage of missing values for all columns in the `combined_df` to understand the current data completeness.



In [ ]:
print("Re-evaluating percentage of missing values for all columns:")

# Calculate the percentage of missing values for each column
missing_percentage_after_cleaning = combined_df.isnull().sum() / len(combined_df) * 100

# Display columns with missing values, sorted in descending order
missing_percentage_after_cleaning = missing_percentage_after_cleaning[missing_percentage_after_cleaning > 0].sort_values(ascending=False)

# Display the results
if not missing_percentage_after_cleaning.empty:
    display(pd.DataFrame(missing_percentage_after_cleaning, columns=['Missing Percentage']))
else:
    print("No missing values found after cleaning and feature engineering.")

## Summary of Data Cleaning and Feature Engineering Steps Performed:

1.  **Numerical Outlier Cleaning (Latitude & Longitude)**:
    *   'Latitude' values outside the range of -90 to 90 were replaced with `pd.NA`.
    *   'Longitude' values outside the range of -180 to 180 were replaced with `pd.NA`.

2.  **Numerical Outlier Cleaning (Injury Counts)**:
    *   For 'FatalInjuryCount', 'SeriousInjuryCount', 'MinorInjuryCount', 'OnboardInjuryCount', and 'OnGroundInjuryCount', any negative values were replaced with `pd.NA` as injury counts cannot be negative.

3.  **Outlier Flagging (`is_extreme_injury`)**:
    *   A new boolean column `is_extreme_injury` was created. This column flags rows where either 'MinorInjuryCount' or 'OnboardInjuryCount' exceeds a threshold of 50, allowing for separate analysis of these extreme cases without altering the original injury count values.

4.  **Categorical Imputation**:
    *   All missing values (`NaN`) in columns with `object` (string) data types were imputed by replacing them with the string 'Unknown'.

5.  **Feature Engineering (`total_injuries`)**:
    *   A new column named `total_injuries` was created, which is the sum of 'FatalInjuryCount', 'SeriousInjuryCount', 'MinorInjuryCount', 'OnboardInjuryCount', and 'OnGroundInjuryCount'. Missing values in these injury columns were treated as 0 for the summation.

6.  **Feature Engineering (`severity_category`)**:
    *   A new categorical column named `severity_category` was created based on the `total_injuries` column. Events were categorized into 'Low' (0 injuries), 'Medium' (1 to 5 injuries), or 'High' (more than 5 injuries).

7.  **DataFrame De-fragmentation**:
    *   The `combined_df` DataFrame was explicitly de-fragmented using `.copy()` to improve performance and prevent `PerformanceWarning` during column additions.

## Summary:

### Data Analysis Key Findings
*   A new boolean column, `is_extreme_injury`, was successfully created to flag rows where `MinorInjuryCount` or `OnboardInjuryCount` exceeded a threshold of 50.
*   Only 1 row in the dataset was identified as having extreme injury counts (either `MinorInjuryCount` or `OnboardInjuryCount` greater than 50).
*   Missing values in all object (categorical) columns were imputed with the string 'Unknown'.
*   A new column `total_injuries` was calculated by summing all types of injury counts, treating any missing injury counts as 0.
*   A `severity_category` column was created based on `total_injuries`, categorizing events into 'Low' (0 injuries), 'Medium' (1 to 5 injuries), and 'High' (more than 5 injuries).
*   After these cleaning and feature engineering steps, no columns were found to have missing values.


In [ ]:
print("Summary statistics for numerical columns:")
display(combined_df.describe())

print("\nSummary statistics for object (non-numerical) columns:")
display(combined_df.describe(include='object'))

In [ ]:
print("Details of the extreme injury case (is_extreme_injury = True):")
extreme_case = combined_df[combined_df['is_extreme_injury'] == True]

display(extreme_case[['MinorInjuryCount', 'OnboardInjuryCount', 'total_injuries', 'Narrative', 'source']])

In [ ]:
print("Summary statistics for numerical columns:")
display(combined_df.describe())

print("\nSummary statistics for object (non-numerical) columns:")
display(combined_df.describe(include='object'))

In [ ]:
print(combined_df.columns.tolist())

In [ ]:
for col, dtype in combined_df.dtypes.items():
    print(f"{col}: {dtype}")

In [ ]:
print("Replacing 'Unknown', 'UNK', and empty strings with NaN in object columns...")

# Get a list of all object (string) columns in the current combined_df
object_columns = combined_df.select_dtypes(include='object').columns

# Define values to replace
values_to_replace = ['Unknown', 'UNK', '']

for col in object_columns:
    # Use .replace() with a list of values to replace with pd.NA
    # The regex=False is important when replacing exact strings to avoid regex interpretation
    combined_df[col] = combined_df[col].replace(values_to_replace, pd.NA)

print("Replacement complete.")

In [ ]:
print("Summary statistics for numerical columns:")
display(combined_df.describe())

print("\nSummary statistics for object (non-numerical) columns:")
display(combined_df.describe(include='object'))

Removing columns that are not neccesarry for our analysis

In [ ]:
columns_to_keep = [
    'ReportID',
    'EventType',
    'City',
    'State',
    'Country',
    'Mode',
    'ReportType',
    'EventDate',
    'N#',
    'SerialNumber',
    'FatalInjuryCount',
    'SeriousInjuryCount',
    'MinorInjuryCount',
    'OnboardInjuryCount',
    'OnGroundInjuryCount',
    'total_injuries',
    'Report 1_Narrative',
    'Report 1.2_Synopsis',
    'Events.3_Detector',
    'Events.4_When Detected',
    'Events.5_Result',
    'Assessments_Contributing Factors / Situations',
    'Assessments.1_Primary Problem'
]

# Filter the DataFrame to keep only the specified columns
combined_df = combined_df[columns_to_keep]

print(f"DataFrame now contains {combined_df.shape[1]} columns.")
print("First 5 rows of the filtered DataFrame:")
display(combined_df.head())
print("\nUpdated column list and their data types:")
combined_df.info()

In [ ]:
combined_df.info()

 identify all columns with 'object' dtype in `combined_df` that contain the string 'Unknown'. This will involve selecting object-type columns and then checking each column's unique values for the presence of 'Unknown'.



In [ ]:
print("Identifying object columns containing 'Unknown' values...")

# Get a list of all object (string) columns
object_columns = combined_df.select_dtypes(include='object').columns

# List to store columns that contain 'Unknown'
columns_with_unknown = []

for col in object_columns:
    if 'Unknown' in combined_df[col].unique():
        columns_with_unknown.append(col)

print(f"Found {len(columns_with_unknown)} object columns containing 'Unknown' values.")

if columns_with_unknown:
    print("\nObject columns containing 'Unknown' values:")
    for col in columns_with_unknown:
        print(f"- {col}")
else:
    print("No object columns found containing 'Unknown' values.")


**Reasoning**:
 calculate the count of 'Unknown' entries for each identified column, sort them in descending order, and then display the top 5 columns along with their unique value counts. This will highlight which columns have the most 'Unknown' values and give a sense of their distribution.



In [ ]:
print("Displaying unique value counts for top 5 columns with the most 'Unknown' entries...")

# Calculate the count of 'Unknown' in each column
unknown_counts = {col: (combined_df[col] == 'Unknown').sum() for col in columns_with_unknown}

# Sort columns by the number of 'Unknown' values in descending order
sorted_columns_by_unknown = sorted(unknown_counts.items(), key=lambda item: item[1], reverse=True)

# Display unique values and their counts for the top 5 columns with the most 'Unknown' entries
num_columns_to_display = 5
for col_name, count in sorted_columns_by_unknown[:num_columns_to_display]:
    print(f"\nColumn: '{col_name}' (Count of 'Unknown': {count})")
    display(combined_df[col_name].value_counts(dropna=False))


# Task
Create a `SubsystemAffected` column as a categorical variable (Engine, Electrical, LandingGear, Avionics, Cabin, Hydraulics, or Other/Unknown) in the `combined_df` DataFrame. The values for this column should be derived by searching for keywords within the 'Report 1_Narrative' and 'Report 1.2_Synopsis' columns. If no specific subsystem is identified, default the value to 'Other/Unknown'.

## Add SubsystemAffected Column

### Subtask:
Create the SubsystemAffected column as a categorical variable (Engine, Electrical, LandingGear, Avionics, Cabin, Hydraulics, or Other/Unknown) in the combined_df DataFrame. The values for this column should be derived by searching for keywords within the 'Report 1_Narrative' and 'Report 1.2_Synopsis' columns. If no specific subsystem is identified, default the value to 'Other/Unknown'.


**Reasoning**:
I need to initialize the new `SubsystemAffected` column in `combined_df` with the default value 'Other/Unknown' for all rows, as per the first instruction of the subtask.



In [ ]:
print("Initializing 'SubsystemAffected' column with 'Other/Unknown'...")
combined_df['SubsystemAffected'] = 'Other/Unknown'
print("Initialization complete.")
print("First 5 rows with 'SubsystemAffected' column:")
display(combined_df[['Report 1_Narrative', 'Report 1.2_Synopsis', 'SubsystemAffected']].head())

**Reasoning**:
As per the instructions, I need to define a dictionary of subsystems and their corresponding keywords to be searched in the narrative columns.



In [ ]:
print("Defining subsystem keywords...")

subsystem_keywords = {
    'Engine': ['engine', 'propeller', 'turbine', 'jet', 'apu', 'motor', 'thrust', 'compressor', 'fan', 'combustor', 'exhaust', 'fuel pump', 'oil pressure', 'rpm', 'cylinder'],
    'Electrical': ['electrical', 'battery', 'generator', 'alternator', 'avionics power', 'circuit breaker', 'bus bar', 'wiring', 'voltage', 'current', 'relay', 'power supply'],
    'LandingGear': ['landing gear', 'wheel', 'tire', 'brake', 'retraction', 'extension', 'strut', 'actuator', 'gear indication', 'nose wheel', 'main gear'],
    'Avionics': ['avionics', 'navigation', 'radio', 'gps', 'autopilot', 'fms', 'instrument', 'display', 'radar', 'transponder', 'communication', 'adi', 'hsi', 'altimeter'],
    'Cabin': ['cabin', 'passenger', 'door', 'emergency exit', 'lavatory', 'galley', 'seat', 'oxygen mask', 'pressurization', 'environmental control', 'cargo hold'],
    'Hydraulics': ['hydraulic', 'fluid', 'pump', 'actuator', 'reservoir', 'pressure', 'line', 'leak', 'accumulator', 'valve', 'power control']
}

print("Subsystem keywords defined.")

**Reasoning**:
Now, I will define a function that iterates through the defined `subsystem_keywords` and checks if any of the keywords for a given subsystem are present in the 'Report 1_Narrative' or 'Report 1.2_Synopsis' columns of a DataFrame row. The function will return the first matching subsystem or 'Other/Unknown' if no match is found, as per the subtask instructions.



In [ ]:
print("Defining function to identify SubsystemAffected...")

def identify_subsystem(row):
    narrative = str(row['Report 1_Narrative']).lower()
    synopsis = str(row['Report 1.2_Synopsis']).lower()

    for subsystem, keywords in subsystem_keywords.items():
        for keyword in keywords:
            if keyword in narrative or keyword in synopsis:
                return subsystem
    return 'Other/Unknown'

print("Function 'identify_subsystem' defined.")

**Reasoning**:
I need to apply the `identify_subsystem` function to each row of the `combined_df` DataFrame to populate the `SubsystemAffected` column, as specified in the subtask instructions. After applying, I will display the value counts of the new column to verify the distribution of categories.



In [ ]:
print("Applying 'identify_subsystem' function to populate 'SubsystemAffected' column...")

# Apply the function to each row to populate the 'SubsystemAffected' column
combined_df['SubsystemAffected'] = combined_df.apply(identify_subsystem, axis=1)

print("'SubsystemAffected' column populated.")

print("\nValue counts for 'SubsystemAffected' column:")
display(combined_df['SubsystemAffected'].value_counts())

## Add PhaseOfFlight Column

### Subtask:
Create the PhaseOfFlight column as a categorical variable (Taxi, Takeoff, Climb, Cruise, Descent, Landing, or Other/Unknown). Values will be assigned by searching for relevant keywords in the narrative columns. If no specific phase is identified, it will default to 'Other/Unknown'.


**Reasoning**:
I need to define the `phase_of_flight_keywords` dictionary as specified in the instructions (step 2) before defining and applying the `identify_phase_of_flight` function.



In [ ]:
print("Defining phase of flight keywords...")

phase_of_flight_keywords = {
    'Taxi': ['taxi', 'pushback', 'gate', 'ramp', 'apron', 'tow'],
    'Takeoff': ['takeoff', 'rotation', 'lift-off', 'departure', 'runway', 'initial climb'],
    'Climb': ['climb', 'ascending', 'enroute climb', 'climbing'],
    'Cruise': ['cruise', 'en route', 'altitude', 'level flight', 'flight level'],
    'Descent': ['descent', 'descending', 'approach preparation', 'top of descent'],
    'Landing': ['landing', 'touchdown', 'final approach', 'flare', 'rollout', 'arrival', 'touch and go'],
    'Go-around': ['go-around', 'rejected landing', 'missed approach'],
    'Holding': ['holding', 'pattern', 'stack']
}

print("Phase of flight keywords defined.")

**Reasoning**:
Now that the `phase_of_flight_keywords` are defined, I need to define the `identify_phase_of_flight` function that will use these keywords to determine the phase of flight from the narrative columns.



# Task
**Reasoning**:
Now that the `phase_of_flight_keywords` are defined, I will define the `identify_phase_of_flight` function. This function will iterate through these keywords and check for their presence in the 'Report 1_Narrative' or 'Report 1.2_Synopsis' columns for each row. If a keyword is found, the corresponding phase of flight will be assigned; otherwise, it will default to 'Other/Unknown'. This function will then be applied to the `combined_df` to create the `PhaseOfFlight` column.



## Define identify_phase_of_flight Function

### Subtask:
Define a function to identify the phase of flight based on keywords in narrative columns.


**Reasoning**:
I need to define the `identify_phase_of_flight` function as per the subtask instructions, which will take a DataFrame row and use the previously defined `phase_of_flight_keywords` to determine the phase of flight.



In [ ]:
print("Defining function to identify PhaseOfFlight...")

def identify_phase_of_flight(row):
    narrative = str(row['Report 1_Narrative']).lower()
    synopsis = str(row['Report 1.2_Synopsis']).lower()

    for phase, keywords in phase_of_flight_keywords.items():
        for keyword in keywords:
            if keyword in narrative or keyword in synopsis:
                return phase
    return 'Other/Unknown'

print("Function 'identify_phase_of_flight' defined.")

**Reasoning**:
The `identify_phase_of_flight` function has been defined, and now I need to apply this function to each row of the `combined_df` DataFrame to create the new `PhaseOfFlight` column. After populating the column, I will display its value counts to confirm the new categories.



In [ ]:
print("Applying 'identify_phase_of_flight' function to populate 'PhaseOfFlight' column...")

# De-fragment the DataFrame to avoid PerformanceWarning before adding a new column
combined_df = combined_df.copy()

# Apply the function to each row to populate the 'PhaseOfFlight' column
combined_df['PhaseOfFlight'] = combined_df.apply(identify_phase_of_flight, axis=1)

print("'PhaseOfFlight' column populated.")

print("\nValue counts for 'PhaseOfFlight' column:")
display(combined_df['PhaseOfFlight'].value_counts())

## Add SensorWarning Column

### Subtask:
Create the SensorWarning column (0 or 1). A value of 1 will be assigned if keywords indicating a sensor warning or alert are found in the narrative columns, or if there are specific entries in 'Events.3_Detector' or 'Events.4_When Detected' that imply a warning. Otherwise, it will be 0.


**Reasoning**:
I need to initialize the new `SensorWarning` column in `combined_df` with a default value of 0, as per the first instruction of the subtask.



In [ ]:
print("Initializing 'SensorWarning' column with 0...")
combined_df['SensorWarning'] = 0
print("Initialization complete.")
print("First 5 rows with 'SensorWarning' column:")
display(combined_df[['Report 1_Narrative', 'Report 1.2_Synopsis', 'SensorWarning']].head())

**Reasoning**:
As per the subtask instructions, I need to define a list of keywords related to sensor warnings and alerts to be used for searching within the narrative columns.



In [ ]:
print("Defining sensor warning keywords...")

sensor_warning_keywords = [
    'warning', 'alert', 'caution', 'alarm', 'light', 'indicator',
    'failure', 'malfunction', 'anomaly', 'detector', 'smoke', 'fire',
    'overheat', 'stall', 'low pressure', 'high temp', 'hot start', 'chip detector',
    'gear warning', 'takeoff warning', 'master warning', 'master caution', 'EICAS', 'ECAM', 'ACARS'
]

print("Sensor warning keywords defined.")

**Reasoning**:
As per instruction 3 of the subtask, I need to define another list of keywords/values to check in the 'Events.3_Detector' and 'Events.4_When Detected' columns that might imply a sensor warning.



In [ ]:
print("Defining detector/when detected keywords...")

detector_when_detected_keywords = [
    'sensor', 'detector', 'ecam', 'eicas', 'acars', 'warning system',
    'master warning', 'master caution', 'alert system', 'indication',
    'light came on', 'aural warning', 'visual alert'
]

print("Detector/when detected keywords defined.")

**Reasoning**:
Now that all keyword lists are defined, I need to create a function that applies the logic described in the subtask instructions (step 4). This function will check the narrative columns ('Report 1_Narrative', 'Report 1.2_Synopsis') and the detector columns ('Events.3_Detector', 'Events.4_When Detected') for the defined keywords, and return 1 if a sensor warning is implied, otherwise 0.



In [ ]:
print("Defining function to identify sensor warnings...")

def identify_sensor_warning(row):
    narrative = str(row['Report 1_Narrative']).lower()
    synopsis = str(row['Report 1.2_Synopsis']).lower()
    detector = str(row['Events.3_Detector']).lower()
    when_detected = str(row['Events.4_When Detected']).lower()

    # Check narrative and synopsis for sensor warning keywords
    for keyword in sensor_warning_keywords:
        if keyword in narrative or keyword in synopsis:
            return 1

    # Check detector and when_detected columns for relevant keywords
    for keyword in detector_when_detected_keywords:
        if keyword in detector or keyword in when_detected:
            return 1

    return 0

print("Function 'identify_sensor_warning' defined.")

**Reasoning**:
Now that the `identify_sensor_warning` function is defined, I need to apply this function to each row of the `combined_df` DataFrame to populate the `SensorWarning` column. After populating, I will display the value counts for the new column to verify the distribution of 0s and 1s, as per the subtask instructions.



In [ ]:
print("Applying 'identify_sensor_warning' function to populate 'SensorWarning' column...")

# De-fragment the DataFrame to avoid PerformanceWarning before adding a new column
combined_df = combined_df.copy()

# Apply the function to each row to populate the 'SensorWarning' column
combined_df['SensorWarning'] = combined_df.apply(identify_sensor_warning, axis=1)

print("'SensorWarning' column populated.")

print("\nValue counts for 'SensorWarning' column:")
display(combined_df['SensorWarning'].value_counts())

## Add HydraulicPressureDrop Column

### Subtask:
Create the `HydraulicPressureDrop` column with values between 0 and 30, synthetically generated based on `SubsystemAffected` and modulated by `total_injuries`.


**Reasoning**:
First, I need to import the `numpy` library as `np` to use its random number generation and clipping functionalities, as specified in the instructions.



In [ ]:
import numpy as np

print("Imported numpy as np.")

**Reasoning**:
As per the instructions, I need to de-fragment the `combined_df` DataFrame by creating a copy to prevent `PerformanceWarning` when adding new columns.



In [ ]:
print("De-fragmenting combined_df...")
combined_df = combined_df.copy()
print("combined_df de-fragmented.")

**Reasoning**:
I need to define the `generate_hydraulic_pressure_drop` function as specified in the instructions (step 3), which will take a DataFrame row as input and calculate a synthetic hydraulic pressure drop value.



In [ ]:
print("Defining function to generate HydraulicPressureDrop...")

def generate_hydraulic_pressure_drop(row):
    # Ensure total_injuries is numeric, handle potential NaNs
    total_injuries = pd.to_numeric(row['total_injuries'], errors='coerce')
    if pd.isna(total_injuries):
        total_injuries = 0 # Default to 0 if total_injuries is NaN

    if row['SubsystemAffected'] == 'Hydraulics':
        base_value = np.random.uniform(low=15, high=28)
    else:
        base_value = np.random.uniform(low=0, high=12)

    # Calculate modulation factor based on total_injuries
    # Add 1 to max() to prevent division by zero and ensure positive factor
    max_total_injuries = combined_df['total_injuries'].max()
    if max_total_injuries == 0:
        modulation_factor = 1 # If no injuries, no modulation
    else:
        modulation_factor = 1 + (total_injuries / (max_total_injuries + 1))

    # Apply modulation and clip the value to the desired range [0, 30]
    calculated_value = base_value * modulation_factor
    return np.clip(calculated_value, 0, 30)

print("Function 'generate_hydraulic_pressure_drop' defined.")

**Reasoning**:
Now that the `generate_hydraulic_pressure_drop` function is defined, I need to apply this function to each row of the `combined_df` DataFrame to create the new `HydraulicPressureDrop` column, as per the subtask instructions (step 4). After applying, I will display the first few rows of the DataFrame showing the new column (step 5) and its descriptive statistics (step 6).



In [ ]:
print("Applying 'generate_hydraulic_pressure_drop' function to populate 'HydraulicPressureDrop' column...")

# Apply the function to each row
combined_df['HydraulicPressureDrop'] = combined_df.apply(generate_hydraulic_pressure_drop, axis=1)

print("'HydraulicPressureDrop' column populated.")

print("\nFirst 5 rows with 'HydraulicPressureDrop' column:")
display(combined_df[['SubsystemAffected', 'total_injuries', 'HydraulicPressureDrop']].head())

print("\nDescriptive statistics for 'HydraulicPressureDrop' column:")
display(combined_df['HydraulicPressureDrop'].describe())

## Add EngineTempDeviation Column

### Subtask:
Create the `EngineTempDeviation` column with values between 0 and 30, synthetically generated based on `SubsystemAffected` and modulated by `total_injuries`.


**Reasoning**:
I need to define the `generate_engine_temp_deviation` function as specified in the instructions (step 1), which will take a DataFrame row as input and calculate a synthetic engine temperature deviation value.



In [ ]:
print("Defining function to generate EngineTempDeviation...")

def generate_engine_temp_deviation(row):
    # Ensure total_injuries is numeric, handle potential NaNs (step 2)
    total_injuries = pd.to_numeric(row['total_injuries'], errors='coerce')
    if pd.isna(total_injuries):
        total_injuries = 0 # Default to 0 if total_injuries is NaN

    # Generate base value based on SubsystemAffected (step 3)
    if row['SubsystemAffected'] == 'Engine':
        base_value = np.random.uniform(low=15, high=28)
    else:
        base_value = np.random.uniform(low=0, high=12)

    # Calculate modulation factor based on total_injuries (step 4)
    # Add 1 to max() to prevent division by zero and ensure positive factor
    max_total_injuries = combined_df['total_injuries'].max()
    if max_total_injuries == 0:
        modulation_factor = 1 # If no injuries, no modulation
    else:
        modulation_factor = 1 + (total_injuries / (max_total_injuries + 1))

    # Apply modulation and clip the value to the desired range [0, 30] (steps 5 and 6)
    calculated_value = base_value * modulation_factor
    return np.clip(calculated_value, 0, 30) # Return clipped value (step 7)

print("Function 'generate_engine_temp_deviation' defined.")

**Reasoning**:
Now that the `generate_engine_temp_deviation` function is defined, I need to apply this function to each row of the `combined_df` DataFrame to create the new `EngineTempDeviation` column, as per the subtask instructions (step 8). After applying, I will display the first few rows of the DataFrame showing the new column (step 9) and its descriptive statistics (step 10).



In [ ]:
print("Applying 'generate_engine_temp_deviation' function to populate 'EngineTempDeviation' column...")

# Apply the function to each row
combined_df['EngineTempDeviation'] = combined_df.apply(generate_engine_temp_deviation, axis=1)

print("'EngineTempDeviation' column populated.")

print("\nFirst 5 rows with 'EngineTempDeviation' column:")
display(combined_df[['SubsystemAffected', 'total_injuries', 'EngineTempDeviation']].head())

print("\nDescriptive statistics for 'EngineTempDeviation' column:")
display(combined_df['EngineTempDeviation'].describe())

## Add ElectricalLoadSpike Column

### Subtask:
Create the `ElectricalLoadSpike` column with values between 0 and 20, synthetically generated based on `SubsystemAffected` and modulated by `total_injuries`. If `SubsystemAffected` is 'Electrical', assign a higher, randomized value within the range; otherwise, assign a lower, randomized value.


**Reasoning**:
I need to define the `generate_electrical_load_spike` function as specified in the instructions (step 1), which will take a DataFrame row as input and calculate a synthetic electrical load spike value.



In [ ]:
print("Defining function to generate ElectricalLoadSpike...")

def generate_electrical_load_spike(row):
    # Ensure total_injuries is numeric, handle potential NaNs (step 2)
    total_injuries = pd.to_numeric(row['total_injuries'], errors='coerce')
    if pd.isna(total_injuries):
        total_injuries = 0 # Default to 0 if total_injuries is NaN

    # Generate base value based on SubsystemAffected (step 3)
    if row['SubsystemAffected'] == 'Electrical':
        base_value = np.random.uniform(low=10, high=18)
    else:
        base_value = np.random.uniform(low=0, high=8)

    # Calculate modulation factor based on total_injuries (step 4)
    # Add 1 to max() to prevent division by zero and ensure positive factor
    max_total_injuries = combined_df['total_injuries'].max()
    if max_total_injuries == 0:
        modulation_factor = 1 # If no injuries, no modulation
    else:
        modulation_factor = 1 + (total_injuries / (max_total_injuries + 1))

    # Apply modulation and clip the value to the desired range [0, 20] (steps 5 and 6)
    calculated_value = base_value * modulation_factor
    return np.clip(calculated_value, 0, 20) # Return clipped value (step 7)

print("Function 'generate_electrical_load_spike' defined.")

**Reasoning**:
Now that the `generate_electrical_load_spike` function is defined, I need to apply this function to each row of the `combined_df` DataFrame to create the new `ElectricalLoadSpike` column, as per the subtask instructions (step 8). After applying, I will display the first few rows of the DataFrame showing the new column (step 9) and its descriptive statistics (step 10).



In [ ]:
print("Applying 'generate_electrical_load_spike' function to populate 'ElectricalLoadSpike' column...")

# Apply the function to each row
combined_df['ElectricalLoadSpike'] = combined_df.apply(generate_electrical_load_spike, axis=1)

print("'ElectricalLoadSpike' column populated.")

print("\nFirst 5 rows with 'ElectricalLoadSpike' column:")
display(combined_df[['SubsystemAffected', 'total_injuries', 'ElectricalLoadSpike']].head())

print("\nDescriptive statistics for 'ElectricalLoadSpike' column:")
display(combined_df['ElectricalLoadSpike'].describe())

## Add RootCauseCategory Column

### Subtask:
Create the `RootCauseCategory` column as a categorical variable (Human, Mechanical, Weather, ATC, Unknown) in the `combined_df` DataFrame. Values will be determined by analyzing keywords in 'Assessments_Contributing Factors / Situations', 'Assessments.1_Primary Problem', and narrative fields. If no clear category is found, it will default to 'Unknown'.


**Reasoning**:
As per the first instruction of the subtask, I need to initialize the new `RootCauseCategory` column in `combined_df` with the default value 'Unknown'.



In [ ]:
print("Initializing 'RootCauseCategory' column with 'Unknown'...")
combined_df['RootCauseCategory'] = 'Unknown'
print("Initialization complete.")
print("First 5 rows with 'RootCauseCategory' column:")
display(combined_df[['Assessments_Contributing Factors / Situations', 'Assessments.1_Primary Problem', 'Report 1_Narrative', 'Report 1.2_Synopsis', 'RootCauseCategory']].head())

**Reasoning**:
As per the second instruction of the subtask, I need to define the `root_cause_keywords` dictionary with categories and their associated keywords.



In [ ]:
print("Defining root cause keywords...")

root_cause_keywords = {
    'Human': ['human error', 'pilot error', 'controller error', 'fatigue', 'distraction', 'procedure violation', 'lack of training', 'crew coordination'],
    'Mechanical': ['mechanical failure', 'engine failure', 'structural failure', 'component failure', 'system malfunction', 'equipment fault', 'maintenance error', 'gear malfunction'],
    'Weather': ['weather', 'wind shear', 'turbulence', 'icing', 'thunderstorm', 'fog', 'visibility', 'crosswind', 'microburst'],
    'ATC': ['air traffic control', 'atc instruction', 'controller action', 'clearance', 'vectoring', 'airspace violation'],
    'Unknown': [] # Default, will be used if no other category matches
}

print("Root cause keywords defined.")

**Reasoning**:
Now that the `root_cause_keywords` dictionary is defined, I need to define the `identify_root_cause` function as specified in the instructions (step 3). This function will process the narrative and assessment columns to determine the root cause category.



In [ ]:
print("Defining function to identify root cause...")

def identify_root_cause(row):
    # Combine relevant text fields for keyword search
    text_to_search = (
        str(row['Assessments_Contributing Factors / Situations']).lower() + " " +
        str(row['Assessments.1_Primary Problem']).lower() + " " +
        str(row['Report 1_Narrative']).lower() + " " +
        str(row['Report 1.2_Synopsis']).lower()
    )

    for category, keywords in root_cause_keywords.items():
        if category == 'Unknown': # Skip the 'Unknown' category itself for keyword matching
            continue
        for keyword in keywords:
            if keyword in text_to_search:
                return category
    return 'Unknown'

print("Function 'identify_root_cause' defined.")

**Reasoning**:
Now that the `identify_root_cause` function is defined, I need to apply this function to each row of the `combined_df` DataFrame to populate the `RootCauseCategory` column, and then display its value counts as per the subtask instructions.



In [ ]:
print("Applying 'identify_root_cause' function to populate 'RootCauseCategory' column...")

# De-fragment the DataFrame to avoid PerformanceWarning before adding a new column
combined_df = combined_df.copy()

# Apply the function to each row to populate the 'RootCauseCategory' column
combined_df['RootCauseCategory'] = combined_df.apply(identify_root_cause, axis=1)

print("'RootCauseCategory' column populated.")

print("\nValue counts for 'RootCauseCategory' column:")
display(combined_df['RootCauseCategory'].value_counts())

## Add SeverityScore Column

### Subtask:
Create the `SeverityScore` column as a numerical value between 1 and 5. This score will be calculated based on a weighted combination of `total_injuries`, `EventType`, and potentially other factors to represent the overall severity of the event. Events with higher `total_injuries` or more severe `EventType` will receive higher scores.


**Reasoning**:
First, I need to initialize the new `SeverityScore` column in `combined_df` with a default numerical value of 1, as per the first instruction of the subtask.



In [ ]:
print("Initializing 'SeverityScore' column with a default value of 1...")
combined_df['SeverityScore'] = 1
print("Initialization complete.")
print("First 5 rows with 'SeverityScore' column:")
display(combined_df[['total_injuries', 'EventType', 'SeverityScore']].head())

**Reasoning**:
As per the subtask instructions, I will define a mapping for `EventType` to assign a base severity score. This mapping will be a dictionary where common event types are assigned a base score, and a default will be used for others.



# Task
## Define EventType Severity Mapping

### Subtask:
Define a dictionary `event_type_severity_map` to assign a base severity score (e.g., 1-5) to different `EventType` categories. This mapping will be used to help calculate the `SeverityScore`.

## Define EventType Severity Mapping

### Subtask:
Define a dictionary `event_type_severity_map` to assign a base severity score (e.g., 1-5) to different `EventType` categories. This mapping will be used to help calculate the `SeverityScore`.


**Reasoning**:
I need to define the `event_type_severity_map` dictionary, including common EventType categories and their corresponding severity scores, as per the subtask instructions.



In [ ]:
print("Defining event_type_severity_map...")

event_type_severity_map = {
    'ACC': 5, # Accident
    'INC': 3, # Incident
    'NMR': 4, # Near Midair Collision
    'OTH': 1, # Other (default low severity)
    'SYF': 4, # System Failure
    'HAL': 3, # Hard Landing
    'DFC': 2, # Damage to Aircraft
    'ALT': 2, # Altimeter Setting Error
    'FOD': 2, # Foreign Object Damage
    'NAV': 3, # Navigation Error
    'FLC': 3, # Flight Control Issue
    'ENG': 4, # Engine Issue
    'LGA': 3  # Landing Gear Anomaly
}

print("event_type_severity_map defined.")
print("Event Type Severity Map:")
display(event_type_severity_map)

**Reasoning**:
Now that the `event_type_severity_map` is defined, I will define a function `calculate_severity_score` that combines the base severity from `EventType` with the `total_injuries` to produce a final `SeverityScore` between 1 and 5, as per the subtask instructions.



In [ ]:
print("Defining function to calculate SeverityScore...")

def calculate_severity_score(row):
    # Get base severity from EventType, default to 1 if not in map
    event_type_base_severity = event_type_severity_map.get(row['EventType'], 1)

    # Ensure total_injuries is numeric, handle potential NaNs
    total_injuries = pd.to_numeric(row['total_injuries'], errors='coerce')
    if pd.isna(total_injuries):
        total_injuries = 0 # Default to 0 if total_injuries is NaN

    # Normalize total_injuries to a small range to avoid excessive influence
    # Max possible total_injuries is 268 based on describe() from earlier, but let's use a dynamic max
    max_total_injuries_in_df = combined_df['total_injuries'].max()
    if max_total_injuries_in_df == 0:
        normalized_injuries = 0
    else:
        normalized_injuries = (total_injuries / max_total_injuries_in_df) * 2 # Scale to max 2 for impact

    # Combine scores. We want a score between 1 and 5.
    # A simple linear combination or weighted average can work.
    # Let's give more weight to event_type_base_severity initially
    combined_score = (event_type_base_severity * 0.7) + (normalized_injuries * 0.3)

    # Clip the score to be between 1 and 5
    return np.clip(combined_score, 1, 5)

print("Function 'calculate_severity_score' defined.")

**Reasoning**:
Now that the `calculate_severity_score` function is defined, I need to apply this function to each row of the `combined_df` DataFrame to populate the `SeverityScore` column, as per the subtask instructions. After applying, I will display the first few rows of the DataFrame showing the new column and its descriptive statistics.



In [ ]:
print("Applying 'calculate_severity_score' function to populate 'SeverityScore' column...")

# De-fragment the DataFrame to avoid PerformanceWarning before adding a new column
combined_df = combined_df.copy()

# Apply the function to each row to populate the 'SeverityScore' column
combined_df['SeverityScore'] = combined_df.apply(calculate_severity_score, axis=1)

print("'SeverityScore' column populated.")

print("\nFirst 5 rows with 'SeverityScore' column:")
display(combined_df[['EventType', 'total_injuries', 'SeverityScore']].head())

print("\nDescriptive statistics for 'SeverityScore' column:")
display(combined_df['SeverityScore'].describe())

## Add IsCriticalEvent Column

### Subtask:
Create the 'IsCriticalEvent' column (Yes or No) based on predefined criteria, including `SeverityScore`, `total_injuries`, and `EventType`.


**Reasoning**:
As per the first instruction of the subtask, I need to initialize the new 'IsCriticalEvent' column in `combined_df` with the default value 'No' for all rows.



In [ ]:
print("Initializing 'IsCriticalEvent' column with 'No'...")
combined_df['IsCriticalEvent'] = 'No'
print("Initialization complete.")
print("First 5 rows with 'IsCriticalEvent' column:")
display(combined_df[['SeverityScore', 'total_injuries', 'EventType', 'IsCriticalEvent']].head())

**Reasoning**:
Now, I need to define the criteria for what constitutes a 'Critical Event' and then apply these criteria to update the `IsCriticalEvent` column in the `combined_df`. This involves checking the `SeverityScore`, `total_injuries`, and `EventType` as per the instructions (step 2 and 3).



In [ ]:
print("Defining and applying criteria for 'IsCriticalEvent'...")

def determine_critical_event(row):
    is_critical_score = row['SeverityScore'] >= 4
    is_critical_injuries = row['total_injuries'] > 5

    # High-severity EventTypes
    high_severity_event_types = ['ACC', 'NMR', 'SYF']
    is_critical_event_type = row['EventType'] in high_severity_event_types

    if is_critical_score or is_critical_injuries or is_critical_event_type:
        return 'Yes'
    return 'No'

# De-fragment the DataFrame to avoid PerformanceWarning before applying
combined_df = combined_df.copy()

# Apply the function to each row to update the 'IsCriticalEvent' column
combined_df['IsCriticalEvent'] = combined_df.apply(determine_critical_event, axis=1)

print("'IsCriticalEvent' column updated based on criteria.")

print("\nFirst 5 rows with updated 'IsCriticalEvent' column:")
display(combined_df[['SeverityScore', 'total_injuries', 'EventType', 'IsCriticalEvent']].head())

print("\nValue counts for 'IsCriticalEvent' column:")
display(combined_df['IsCriticalEvent'].value_counts())

**Reasoning**:
The previous code failed because `row['EventType']` could be `pd.NA`, leading to a `TypeError` when checked for membership in `high_severity_event_types`. To fix this, I will explicitly check if `EventType` is `pd.NA` and treat it as `False` for the critical event condition; otherwise, I will proceed with the membership check.



In [ ]:
print("Defining and applying criteria for 'IsCriticalEvent'...")

def determine_critical_event(row):
    is_critical_score = row['SeverityScore'] >= 4
    is_critical_injuries = row['total_injuries'] > 5

    # High-severity EventTypes
    high_severity_event_types = ['ACC', 'NMR', 'SYF']

    # Handle potential pd.NA in 'EventType'
    event_type = row['EventType']
    if pd.isna(event_type):
        is_critical_event_type = False
    else:
        is_critical_event_type = event_type in high_severity_event_types

    if is_critical_score or is_critical_injuries or is_critical_event_type:
        return 'Yes'
    return 'No'

# De-fragment the DataFrame to avoid PerformanceWarning before applying
# This line was already executed and dataframe copied, but kept for robustness if cell is run independently
combined_df = combined_df.copy()

# Apply the function to each row to update the 'IsCriticalEvent' column
combined_df['IsCriticalEvent'] = combined_df.apply(determine_critical_event, axis=1)

print("'IsCriticalEvent' column updated based on criteria.")

print("\nFirst 5 rows with updated 'IsCriticalEvent' column:")
display(combined_df[['SeverityScore', 'total_injuries', 'EventType', 'IsCriticalEvent']].head())

print("\nValue counts for 'IsCriticalEvent' column:")
display(combined_df['IsCriticalEvent'].value_counts())

## Summarize New Columns

### Subtask:
Display the first few rows of the DataFrame with the newly added columns to verify their creation and provide summary statistics (head, info, describe) for these new columns.


**Reasoning**:
First, I will display the first 5 rows of the `combined_df` DataFrame, explicitly showing the newly created columns: `SubsystemAffected`, `PhaseOfFlight`, `SensorWarning`, `HydraulicPressureDrop`, `EngineTempDeviation`, `ElectricalLoadSpike`, `RootCauseCategory`, `SeverityScore`, and `IsCriticalEvent` to verify their creation.



In [ ]:
print("Displaying the first 5 rows with the newly added columns:")
display(combined_df[[
    'SubsystemAffected', 'PhaseOfFlight', 'SensorWarning',
    'HydraulicPressureDrop', 'EngineTempDeviation', 'ElectricalLoadSpike',
    'RootCauseCategory', 'SeverityScore', 'IsCriticalEvent'
]].head())

**Reasoning**:
Now, I will print the information about the `combined_df` DataFrame using the `.info()` method to show data types and non-null counts for all columns, as requested by the subtask.



In [ ]:
print("\nInformation about the combined_df DataFrame:")
combined_df.info()

**Reasoning**:
Now, I will display descriptive statistics for all columns in the `combined_df` DataFrame using the `.describe(include='all')` method to include both numerical and object type columns, as requested by the subtask.



In [ ]:
print("\nDescriptive statistics for all columns:")
display(combined_df.describe(include='all'))

## Final Task

### Subtask:
Summarize all new columns added and their derivation methods, along with any insights gained from their initial review.


## Summary:

### Q&A
1.  **How was the `event_type_severity_map` defined?**
    The `event_type_severity_map` was defined as a dictionary assigning integer severity scores (ranging from 1 to 5) to various `EventType` abbreviations. For instance, 'ACC' (Accident) was assigned a severity of 5, 'INC' (Incident) a severity of 3, and 'OTH' (Other) a severity of 1.

2.  **How was the `SeverityScore` calculated?**
    The `SeverityScore` was calculated using a custom function that combined the base severity from the `event_type_severity_map` (weighted 70%) with a normalized value of `total_injuries` (weighted 30%). The `total_injuries` were normalized by scaling them against the maximum injuries observed in the dataset. The final score was then clipped to be between 1 and 5.

3.  **How was the `IsCriticalEvent` column determined?**
    The `IsCriticalEvent` column was determined by evaluating three criteria:
    *   `SeverityScore` greater than or equal to 4.
    *   `total_injuries` greater than 5.
    *   `EventType` being one of 'ACC', 'NMR', or 'SYF' (Accident, Near Midair Collision, System Failure).
    If any of these conditions were met, `IsCriticalEvent` was set to 'Yes'; otherwise, it was 'No'. The calculation also robustly handled potential missing `EventType` values.

### Data Analysis Key Findings
*   The `event_type_severity_map` successfully assigns base severity scores to 13 different `EventType` categories, with 'ACC' having the highest base severity of 5 and 'OTH' having the lowest of 1.
*   The calculated `SeverityScore` ranges from 1.0 to 4.1, with a mean of approximately 1.40, indicating that most events tend to have lower severity scores.
*   The `IsCriticalEvent` column classifies 7,381 events as 'Yes' (critical) and 38,736 events as 'No' (non-critical) out of 46,117 total events.
*   All newly added columns (`SubsystemAffected`, `PhaseOfFlight`, `SensorWarning`, `HydraulicPressureDrop`, `EngineTempDeviation`, `ElectricalLoadSpike`, `RootCauseCategory`, `SeverityScore`, `IsCriticalEvent`) have been successfully created and contain 46,117 non-null entries, ensuring no missing data in these new features.
*   The `SensorWarning` column is a binary flag (0 or 1) with a mean of approximately 0.68, suggesting about 68% of events are associated with a sensor warning.
*   Categorical columns like `SubsystemAffected`, `PhaseOfFlight`, and `RootCauseCategory` frequently show 'Other/Unknown', 'Taxi', and 'Unknown' respectively as their most common categories.

### Insights or Next Steps
*   The predominance of 'Unknown' or 'Other/Unknown' categories in key descriptive columns like `RootCauseCategory`, `SubsystemAffected`, and `PhaseOfFlight` suggests that further data collection or categorization efforts could greatly enhance the granularity and utility of these features for more precise analysis.
*   Given the successful creation of `SeverityScore` and `IsCriticalEvent`, these new metrics can now be leveraged as primary targets or predictors for advanced analytics, such as identifying risk factors, predicting future critical events, or optimizing resource allocation for safety improvements.


In [ ]:
print("Value counts for 'SubsystemAffected' column:")
display(combined_df['SubsystemAffected'].value_counts(dropna=False))

In [ ]:
print("Non-null counts for 'Report 1_Narrative' and 'Report 1.2_Synopsis':")
print(combined_df[['Report 1_Narrative', 'Report 1.2_Synopsis']].count())


In [ ]:
print("Value counts for 'IsCriticalEvent' column:")
display(combined_df['IsCriticalEvent'].value_counts(dropna=False))

In [ ]:
print("Value counts for 'SeverityScore' column:")
display(combined_df['SeverityScore'].value_counts(dropna=False))

In [ ]:
print("\nInformation about the combined_df DataFrame:")
combined_df.info()

In [ ]:
print("\nDescriptive statistics for all columns:")
display(combined_df.describe(include='all'))

Saving the combined dataframe into a csv file to be loaded into the sql server for further processing


In [ ]:
print("Saving 'combined_df' to 'source_file.csv'...")
combined_df.to_csv('source_file.csv', index=False)
print("'combined_df' successfully saved to 'source_file.csv'.")

In [ ]:
from google.colab import files

print("Downloading 'source_file.csv'...")
files.download('source_file.csv')
print("Download initiated. Please check your browser's download folder.")